File: VerifySignature.java
/**
* Project: Final Project - Digital Signature Standard
* Created By: Alex Cadigan, Jacob Scott, Skyler Norgaard
* Date Created: 3/11/2017
*/

// Imports:

import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.SecureRandom;
import java.security.Security;
import java.security.Signature;

/**
* This class contains code to verify a signature using the Digital Signature Standard algorithm.
*/
public class VerifySignature
{

	/**
	* Runs the Digital Signature Standard algorithm to verify a signature.
	*/
	public void verifySignature()
	{
		Security.addProvider(new org.bouncycastle.jce.provider.BouncyCastleProvider());
	}
	

// 	import java.security.KeyPair;
// import java.security.KeyPairGenerator;
// import java.security.PrivateKey;
// import java.security.PublicKey;
// import java.security.SecureRandom;
// import java.security.Security;
// import java.security.Signature;

// public class MainClass {
//   public static void main(String args[]) throws Exception {
//     Security.addProvider(new org.bouncycastle.jce.provider.BouncyCastleProvider());

//     KeyPair keyPair = generateKeyPair(999);

//     byte[] data = { 65, 66, 67, 68, 69, 70, 71, 72, 73, 74 };
//     byte[] digitalSignature = signData(data, keyPair.getPrivate());

//     boolean verified;

//     verified = verifySig(data, keyPair.getPublic(), digitalSignature);
//     System.out.println(verified) ;

//     keyPair = generateKeyPair(888);
//     verified = verifySig(data, keyPair.getPublic(), digitalSignature);
//     System.out.println(verified);

//   }

//   public static byte[] signData(byte[] data, PrivateKey key) throws Exception {
//     Signature signer = Signature.getInstance("SHA1withDSA");
//     signer.initSign(key);
//     signer.update(data);
//     return (signer.sign());
//   }

//   public static boolean verifySig(byte[] data, PublicKey key, byte[] sig) throws Exception {
//     Signature signer = Signature.getInstance("SHA1withDSA");
//     signer.initVerify(key);
//     signer.update(data);
//     return (signer.verify(sig));

//   }

//   public static KeyPair generateKeyPair(long seed) throws Exception {
//     Security.addProvider(new org.bouncycastle.jce.provider.BouncyCastleProvider());
//     KeyPairGenerator keyGenerator = KeyPairGenerator.getInstance("DSA");
//     SecureRandom rng = SecureRandom.getInstance("SHA1PRNG", "SUN");
//     rng.setSeed(seed);
//     keyGenerator.initialize(1024, rng);

//     return (keyGenerator.generateKeyPair());
//   }
	//}
}

File: VerifySignature.py
'''
Project: Final Project - Digital Signature Standard
Created By: Skyler Norgaard, Jacob Scott, and Alex Cadigan
Date Created: 3/12/2017

Description: This file contains code to verify a signature using a Digital Signature Standard Algorithm.
'''

# Imports:

import Helper

# Initiates the algorithm to verify a signature
def verifySignature():

	# Gets the signed message from the user
	message = input("\nEnter the signed message you have received:\t")

	# Gets the signer's public value 'p'
	p = input("\nEnter the signer's public value 'p':\t")
	while (not Helper.isPositiveNumber(p)):

		p = input("\nInvalid entry!  Enter the signer's public value 'p':\t")

	# Gets the signer's public value 'q'
	q = input("\nEnter the signer's public value 'q':\t")
	while (not Helper.isPositiveNumber(q)):

		q = input("\nInvalid entry!  Enter the signer's public value 'q':\t")

	# Gets the signer's public value 'g'
	g = input("\nEnter the signer's public value 'g':\t")
	while (not Helper.isPositiveNumber(g)):

		g = input("\nInvalid entry!  Enter the signer's public value 'g':\t")

	# Gets the signer's public key
	y = input("\nEnter the signer's public key 'y':\t")
	while (not Helper.isPositiveNumber(y)):

		y = input("\nInvalid entry!  Enter the signer's public key 'y':\t")

	# Gets the signed value 'r' from the user
	r = input("\nEnter the signature value 'r' you have received:\t")
	while (not Helper.isPositiveNumber(r)):

		r = input("\nInvalid entry!  Enter the signature value 'r' you have received:\t")

	# Gets the signed value 's' from the user
	s = input("\nEnter the signature value 's' you have received:\t")
	while (not Helper.isPositiveNumber(s)):

		s = input("\nInvalid entry!  Enter the signature value 's' you have received:\t")

	# Calculates the value 'w'
	sInverse = Helper.calculateModularMultiplicativeInverse(int(s), int(q))
	w = Helper.calculateW(int(q), sInverse)

	# Determines if the user wants to enter a hash value, or use their own
	decision = input("\nWould you like to enter your own hash value (enter 'Yes' or 'No')?\t").upper()
	while (decision != "YES" and decision != "NO"):

		decision = input("\nInvalid entry!  Would you like to enter your own hash value (enter 'Yes' or 'No')?\t").upper()
	
	# Gets the hash value
	if (decision == "YES"):

		H = input("\nEnter your own hash value:\t")
		while (not Helper.isPositiveNumber(H)):

			H = input("\nEntry invalid!  Enter your own hash value:\t")

		H = int(H)

	else:

		H = Helper.secureHashAlgorithm1(message)

	# Calculates the value 'u1'
	u1 = Helper.calculateU1(int(q), w, H)

	# Calculates the value 'u2'
	u2 = Helper.calculateU2(int(q), int(r), w)

	# Calculates the value 'v'
	v = Helper.calculateV(int(p), int(q), int(g), int(y), u1, u2)

	# Tests if the signature is valid
	if (v == int(r)):

		print("\nSuccess!  This signature is valid!")

	else:

		print("\nError!  This signature is not valid!")
